### PR TITLE
rework EML

### LINKED ISSUE #33: Add attachments
### Please describe the feature you would like to see implemented.

At the moment Mail attachments won't be listed and are missing if you resend an email.
This would be awesome!

### LINKED ISSUE #35: rework eml
### Please describe the feature you would like to see implemented.

- [ ] save EML gzipped into blob-field (there are also attachments saved)
- [ ] extract information from EML or using it 1:1 in [MailResendController](https://github.com/FriendsOfShopware/FroshPlatformMailArchive/blob/0.3/src/Controller/Api/MailResendController.php) instead of fields

### PR DESCRIPTION
**Pull request by tinect**
Closes #35
Closes #33

- [x] gzip `eml` file
- [x] test mail-migration
- [x] subscribe to entity deletion to remove files from filesystem, too
- [x] add info about not migrated mails when a user logs in
- [x] AsyncMessageInterface in message

### CODE REVIEW

File: src/Migration/Migration1690743548AddEmlPath.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public function getCreationTimestamp(): int
+    {
+        return 1690743548;
+    }
+
+    public function update(Connection $connection): void
+    {
+        $connection->executeStatement('ALTER TABLE `frosh_mail_archive`
+                                            ADD `eml_path` varchar(2048) NULL;');
```
Comment (shyim) [inline src/Migration/Migration1690743548AddEmlPath.php:18]: why so big that string?

File: src/Resources/app/administration/src/module/frosh-mail-archive/page/frosh-mail-archive-detail/index.js [hunk 2]
```diff
@@ -73,14 +75,25 @@ Component.register('frosh-mail-archive-detail', {
...
-                this.isLoading = false;
-                this.isSuccessful = false;
+                this.resendIsLoading = false;
+                this.resendIsSuccessful = false;
+            });
+        },
+        downloadMail() {
+            this.downloadIsLoading = true;
+
+            this.froshMailArchiveService.downloadMail(this.archive.id).then(() => {
```
Comment (shyim) [inline src/Resources/app/administration/src/module/frosh-mail-archive/page/frosh-mail-archive-detail/index.js:91]: just make it async and await? looks much more clean to me.

File: src/Controller/Api/MailArchiveController.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+{
+    public function __construct(
+        private readonly EntityRepository $froshMailArchiveRepository,
+        private readonly MailSender $mailSender,
+        private readonly RequestStack $requestStack,
+        private readonly EmlFileManager $emlFileManager
+    ) {
+    }
+
+    #[Route(path: '/api/_action/frosh-mail-archive/resend-mail', defaults: ['_routeScope' => ['api']])]
```
Comment (shyim) [inline src/Controller/Api/MailArchiveController.php:34]: move it to the class as once (routeScope).

File: src/Controller/Api/MailArchiveController.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Email;
+use Symfony\Component\Routing\Annotation\Route;
+use ZBateson\MailMimeParser\Header\AddressHeader;
+use ZBateson\MailMimeParser\Header\DateHeader;
+use ZBateson\MailMimeParser\Header\IHeader;
+use ZBateson\MailMimeParser\Header\Part\AddressPart;
+
+class MailArchiveController extends AbstractController
```
Comment (shyim) [inline src/Controller/Api/MailArchiveController.php:24]: We should avoid using AbstractController as a base class for controllers. We don't need something special from it.

File: src/Controller/Api/MailArchiveController.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        $mailId = $request->request->get('mailId');
+
+        if (!\is_string($mailId)) {
+            throw new \RuntimeException('mailId not given');
+        }
+
+        $mailArchive = $this->froshMailArchiveRepository->search(new Criteria([$mailId]), Context::createDefaultContext())->first();
+        if (!$mailArchive instanceof MailArchiveEntity) {
+            throw new \RuntimeException('Cannot find mail in archive');
```
Comment (shyim) [inline src/Controller/Api/MailArchiveController.php:45]: we should throw a ShopwareHttpException so it's properly rendered

File: src/Resources/config/services.xml [hunk 6]
```diff
@@ -21,27 +22,6 @@
 
         <service id="Frosh\MailArchive\Services\MailSender" public="true" decorates="Shopware\Core\Content\Mail\Service\MailSender">
```
Comment (shyim) [inline src/Resources/config/services.xml:23]: autowire 👍

File: src/Resources/public/administration/css/frosh-platform-mail-archive.css [hunk 7]
```diff
@@ -1 +1 @@
-.sw-sidebar-item__scrollable-container{padding:10px}.frosh-mail-archive-detail iframe{width:100%;height:500px;background-color:#fff}.frosh-mail-archive-detail pre{white-space:pre-line}
\ No newline at end of file
+.sw-sidebar-item__scrollable-container{padding:10px}.frosh-mail-archive-detail iframe{width:100%;height:500px;background-color:#fff}.frosh-mail-archive-detail pre{white-space:pre-line}
```
Comment (shyim) [inline src/Resources/public/administration/css/frosh-platform-mail-archive.css:1]: why do we commit this? 😂

File: src/Subscriber/MailArchiveDeleteSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    public function beforeDelete(BeforeDeleteEvent $event): void
+    {
+        /** @var list<string> $affected */
+        $ids = array_values($event->getIds(MailArchiveDefinition::ENTITY_NAME));
+        if (empty($ids)) {
+            return;
+        }
+
+        $criteria = new Criteria($ids);
```
Comment (shyim) [inline src/Subscriber/MailArchiveDeleteSubscriber.php:36]: $criteria->addFields(['emlPath]);

save some traffic

File: src/Task/CleanupTaskHandler.php [hunk 9]
```diff
@@ -33,14 +36,35 @@ public function run(): void
...
+            return;
+        }
+
+        foreach ($result as $item) {
+            $this->emlFileManager->deleteEmlFile($item['eml_path']);
+        }
+
+        $deleteQuery = $this->connection->createQueryBuilder();
+        $deleteQuery->delete(MailArchiveDefinition::ENTITY_NAME);
+        $deleteQuery->where(
```
Comment (shyim) [inline src/Task/CleanupTaskHandler.php:61]: `id IN (:ids)`

->setParameter(....

File: src/Subscriber/MailArchiveDeleteSubscriber.php (hunk 8 above)
Comment (tinect) [reply src/Subscriber/MailArchiveDeleteSubscriber.php:36]: Why is this marked `@internal`?

Comment (tinect) [reply src/Subscriber/MailArchiveDeleteSubscriber.php:36]: ` "Attempted to call an undefined method named \u0022getEmlPath\u0022 of class \u0022Shopware\\Core\\Framework\\DataAbstractionLayer\\PartialEntity\u0022.",`

:-(

**Review by shyim — APPROVED** [review]

File: src/Command/MigrateMailCommand.php [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepository;
+use Symfony\Component\Console\Attribute\AsCommand;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+#[AsCommand('frosh-mailarchive:migrate-mails', 'Migrate mails from database to private filesystem')]
```
Comment (stefanpoensgen) [inline src/Command/MigrateMailCommand.php:17]: frosh:mailarchive:migrate-mails

File: src/Resources/config/services.xml [hunk 11]
```diff
@@ -5,7 +5,8 @@
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
 
     <services>
-        <defaults autowire="true"/>
+        <defaults autowire="true" autoconfigure="true"/>
+        <prototype namespace="Frosh\MailArchive\" resource="../../" />
 
         <service id="Frosh\MailArchive\Content\MailArchive\MailArchiveDefinition">
             <tag name="shopware.entity.definition" entity="frosh_mail_archive"/>
```
Comment (stefanpoensgen) [inline src/Resources/config/services.xml:12]: shopware.entity.definition and shopware.entity.extension could be autowired aswell

Comment (tinect) [reply src/Resources/config/services.xml]: Thank you! 👍

File: src/Command/MigrateMailCommand.php (hunk 10 above)
Comment (tinect) [reply src/Command/MigrateMailCommand.php]: niceeee, yes!

File: src/Content/MailArchive/MailArchiveAttachmentDefinition.php [hunk 12]
```diff
@@ -11,7 +11,9 @@
 use Shopware\Core\Framework\DataAbstractionLayer\Field\ManyToOneAssociationField;
 use Shopware\Core\Framework\DataAbstractionLayer\Field\StringField;
 use Shopware\Core\Framework\DataAbstractionLayer\FieldCollection;
+use Symfony\Component\DependencyInjection\Attribute\AutoconfigureTag;
 
+#[AutoconfigureTag('shopware.entity.definition', attributes: ['entityName' => self::ENTITY_NAME])]
```
Comment (shyim) [inline src/Content/MailArchive/MailArchiveAttachmentDefinition.php:16]: you don't need entityName

File: src/Content/MailArchive/MailArchiveAttachmentCollection.php [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?php declare(strict_types=1);
+
+namespace Frosh\MailArchive\Content\MailArchive;
+
+use Shopware\Core\Framework\DataAbstractionLayer\EntityCollection;
+
+/**
+ * @extends EntityCollection<MailArchiveAttachmentEntity>
+ */
+class MailArchiveAttachmentCollection extends EntityCollection
```
Comment (shyim) [inline src/Content/MailArchive/MailArchiveAttachmentCollection.php:10]: what about we don't create collections and use the already existing EntityCollection?

Comment (tinect) [reply src/Content/MailArchive/MailArchiveAttachmentCollection.php:10]: Uhhhh, this sounds really nice!

**Review by shyim — APPROVED** [review]

**Pull request merged**